core/xim/pm/XPackage.lua

```diff
@@ -9,8 +9,8 @@ local os_info = utils.os_info()
 function new(pkg)
     local instance = {}
     debug.setmetatable(instance, XPackage)
-    instance.pdata = pkg.data.package
-    instance.version = pkg.version
+    instance.version, instance.pdata = _dereference(pkg.version, pkg.data.package)
```

Comment:
If version is a factor affecting independence, so for others. It's a coarse separation here.

Reply:
Here, design an input list to restrict a sibling of a package. Rather directly segregate version and package body.
Suggest an improvement of `XPackage`.
We could transport `PMService` functionalities to `XPackage` itself. There's no need of a factor-style design here.

Reply:
Suppose args list as a restriction to find a sibling (we give the freedom of choosing version, vendors...etc for a package).
```lua
-- args is a predefined config.
function new(args,pkg)
       -- same...
       -- search loigc by args
       ins.data = search_data
       -- possible other thing.
       -- caveat: we shouldn't take accessory such as _map_pkgname etc...
```

Reply:
Transport the logic of `PMService`:
```lua
function XPackage:executor(...)
      -- impl logic
end
```